src/scheduler.rs

```diff
@@ -0,0 +1,378 @@
+        let id = parts.get(3).unwrap_or(&"0").parse().unwrap_or(0); // FIXME: How should we actually handle a bad identifier string?
+        Ok(Self { priority, name, id })
+    }
+}
+
+/// Implementers of `Executable` acts as a callback which executes the job associated with a task.
+trait Executable<R: Resource> {
+    /// `execute` executes a task's job. `preempt.preempt_now()` should be polled as appropriate,
+    /// and execution should terminate if it returns true. Tasks which are no preemptible need not
+    /// ever check for preemption.
+    fn execute(&self, preempt: &dyn Preemption<R>);
+
+    /// Returns true if the job associated with this `Executable` can be preempted. `Executable`s
+    /// which return `true` should periodically poll for preemption while executing.
+    fn is_preemptible(&self) -> bool;
+}
+
+#[derive(Clone)]
+pub struct Task<'a, R: Resource> {
+    /// These are the resources for which the `Task` has been requested to be scheduled,
+    /// in order of preference. It is guaranteed that the `Task` will be scheduled on only one of these.
+    executable: &'a dyn Executable<R>,
+}
+
+pub struct FSScheduler<'a, R: Resource> {
+    root: PathBuf,
+    /// A given `Task` (identified uniquely by a `TaskIdent`) may have multiple `TaskFile`s associated,
+    /// one per `Resource` for which it is currently scheduled (but only one `Resource` will eventually be assigned).
+    task_files: HashMap<TaskIdent, HashSet<TaskFile>>,
+    /// Each `Task` (identified uniquely by a `TaskIdent`) is protected by a `Mutex`.
+    own_tasks: HashMap<TaskIdent, Mutex<&'a Task<'a, R>>>,
+    ident_counter: usize,
+    _r: PhantomData<R>,
+}
+
+impl<'a, R: Resource> FSScheduler<'a, R> {
+    pub fn new(root: PathBuf) -> Result<Self, Error> {
+        create_dir_all(&root)?;
+        Ok(Self {
+            root,
+            task_files: Default::default(),
+            own_tasks: Default::default(),
+            _r: PhantomData,
+            ident_counter: 0,
+        })
+    }
+    pub fn new_ident(&mut self, priority: Priority, name: String) -> TaskIdent {
+        let id = self.ident_counter;
+        self.ident_counter += 1;
+        TaskIdent { priority, name, id }
+    }
+    pub fn schedule(
+        &mut self,
+        task_ident: TaskIdent,
+        task: &'a Task<'a, R>,
+        resources: &[R],
+    ) -> Result<(), Error> {
+        for resource in resources.iter() {
+            let dir = self.root.join(resource.dir_id());
+            create_dir_all(&dir)?;
+            let task_file = task_ident.enqueue_in_dir(&dir)?;
+            self.task_files
+                .entry(task_ident.clone())
+                .or_insert(Default::default())
+                .insert(task_file);
+            self.own_tasks.insert(task_ident.clone(), Mutex::new(&task));
+        }
+        Ok(())
+    }
+}
+
+pub struct FSResourceScheduler<'a, R: Resource> {
+    root_scheduler: FSScheduler<'a, R>,
```

Comment:
Shouldn't this be a reference? Maybe: `Rc<FSScheduler<'a, R>>`
My understanding is that we will have a single root FSScheduler and a few FSResourceSchedulers (One per resource) that point to the root scheduler

Reply:
Yes, just hadn't gotten there yet. It ended up being `Rc<RefCell<…>>`.